gslib/context_config.py

```diff
@@ -0,0 +1,203 @@
+                                             'use_client_certificate',
+                                             None) == 'True'
+
+    atexit.register(self._UnprovisionClientCert)
+
+    if self.use_client_certificate:
+      command_string = config.get('Credentials', 'cert_provider_command', None)
+      if not command_string:
+        raise CertProvisionError('No cert provider detected.')
+
+      self.client_cert_path = os.path.join(gslib.GSUTIL_DIR, 'caa_cert.pem')
+      try:
+        # Certs provisioned using endpoint verification are stored as a
+        # single file holding both the public certificate and the private key.
+        self._ProvisionClientCert(command_string, self.client_cert_path)
+      except CertProvisionError as e:
+        self.logger.error('Failed to provision client certificate: %s' % e)
+
+  def _ProvisionClientCert(self, command_string, cert_path):
+    """Executes certificate provider to obtain client certificate and keys."""
+    # Monkey-patch command line args to get password-protected certificate.
+    # Adds password flag if it's not already there.
+    password_arg = ' --with_passphrase'
+    if ('--print_certificate' in command_string and
+        password_arg not in command_string):
+      command_string += password_arg
+
+    try:
+      command_process = subprocess.Popen(command_string.split(' '),
+                                         stdout=subprocess.PIPE,
+                                         stderr=subprocess.PIPE)
+
+      command_stdout, command_stderr = command_process.communicate()
+      if command_process.returncode != 0:
+        raise CertProvisionError(command_stderr)
+
+      sections = _SplitPemIntoSections(command_stdout, self.logger)
+      with open(cert_path, 'w+') as f:
+        f.write(sections['CERTIFICATE'])
+        f.write(sections['ENCRYPTED PRIVATE KEY'])
+      self.client_cert_password = sections['PASSPHRASE'].splitlines()[1]
+    except OSError as e:
+      raise CertProvisionError(e)
+    except KeyError as e:
+      raise CertProvisionError(
+          'Invalid output format from certificate provider, no %s' % e)
+
+  def _UnprovisionClientCert(self):
+    """Cleans up any files or resources provisioned during config init."""
+    if self.client_cert_path is not None:
+      try:
+        os.remove(self.client_cert_path)
+        self.logger.debug('Unprovisioned client cert: %s' %
+                          self.client_cert_path)
+      except OSError as e:
+        self.logger.error('Failed to remove client certificate: %s' % e)
+
+
+def create_context_config(logger):
+  """Should be run once at gsutil startup. Creates global singleton.
+
+  Args:
+    logger (logging.logger): For logging during config functions.
+
+  Returns:
+    New ContextConfig singleton or None if singleton already exists.
+  """
+  global _singleton_config
+  if not _singleton_config:
+    _singleton_config = _ContextConfig(logger)
+    return _singleton_config
+  return None
```

Comment:
I think raising and error if the function get's called twice might be better.

Reply:
Done